README.md
<h1><a href="https://ergo.services"><img src=".github/images/logo.svg" alt="Ergo Framework" width="159" height="49"></a></h1>

<!--[![Gitbook Documentation](https://img.shields.io/badge/GitBook-Documentation-f37f40?style=plastic&logo=gitbook&logoColor=white&style=flat)](https://docs.ergo.services) -->
[![GoDoc](https://pkg.go.dev/badge/ergo-services/ergo)](https://pkg.go.dev/github.com/sllt/ergo)
[![MIT license](https://img.shields.io/badge/license-MIT-brightgreen.svg)](https://opensource.org/licenses/MIT)
[![Telegram Community](https://img.shields.io/badge/Telegram-Community-blue?style=flat&logo=telegram)](https://t.me/ergo_services)
[![Discord Community](https://img.shields.io/badge/Discord-Community-5865F2?style=flat&logo=discord&logoColor=white)](https://discord.gg/sdscxKGV62)
[![Twitter](https://img.shields.io/badge/Twitter-ergo__services-1DA1F2?style=flat&logo=twitter&logoColor=white)](https://twitter.com/ergo_services)

Technologies and design patterns of Erlang/OTP have been proven over the years. Now in Golang.
Up to x5 times faster than original Erlang/OTP in terms of network messaging.
The easiest way to create an OTP-designed application in Golang.

[https://ergo.services](https://ergo.services)

### Purpose ###

The goal of this project is to leverage Erlang/OTP experience with Golang performance. Ergo Framework implements [DIST protocol](https://erlang.org/doc/apps/erts/erl_dist_protocol.html), [ETF data format](https://erlang.org/doc/apps/erts/erl_ext_dist.html) and [OTP design patterns](https://erlang.org/doc/design_principles/des_princ.html) `gen.Server`, `gen.Supervisor`, `gen.Application` which makes you able to create distributed, high performance and reliable microservice solutions having native integration with Erlang infrastructure

### Cloud ###

Distributed Cloud is coming. With Ergo Framework you can join your services into a single cluster with transparent networking using our **Cloud Overlay Network** where they can connect to each other smoothly, no matter where they run - AWS, Azure or GCP, or anywhere else. All these connections are secured with end-to-end encryption. Read more in this article [https://blog.ergo.services/cloud-overlay-network-3a133d47efe5](https://blog.ergo.services/cloud-overlay-network-3a133d47efe5).

### Features ###

![image](https://user-images.githubusercontent.com/118860/113710255-c57d5500-96e3-11eb-9970-20f49008a990.png)

* Support Erlang 25 - allows you connect your node to (and accept connection from) any Erlang/Elixir node within a cluster
* Spawn Erlang-like processes
* Register/unregister processes with simple atom
* Set of ready-to-use disign patterns (behaviors)
  * `gen.Server` behavior with atomic state and Erlang's gen_server support to make sync request `ServerProcess.Call`, async - `ServerProcess.Cast` or `Process.Send` in fashion of `gen_server:call`, `gen_server:cast`, `erlang:send` accordingly
  * `gen.Supervisor` behavior with all known [restart strategies](https://erlang.org/doc/design_principles/sup_princ.html#restart-strategy) (One For One, One For All, Rest For One, Simple One For One)
  * `gen.Application` behavior with all known [starting types](https://erlang.org/doc/design_principles/applications.html#application-start-types) (Permanent, Temporary, Transient)
  * `gen.Pool` a basic design pattern with a pool of workers. All messages/requests received by the pool process are forwarded to the workers using the "Round Robin" algorithm. The worker process is automatically restarting on termination
  * `gen.TCP` - socket acceptor pool for TCP protocols. This behavior aims to provide everything you need to accept TCP connections and process packets with a small code base and low latency while being easy to use.
  * `gen.UDP` - acceptor pool for UDP protocols. This behavior provides the same feature set as TCP but for handling UDP packets using pool of handlers
  * `gen.Web` - Web API Gateway behavior. This behavior allows you to listen HTTP port and handle HTTP-request using pool of workers.
  * `gen.Stage` behavior support (originated from Elixir's [GenStage](https://hexdocs.pm/gen_stage/GenStage.html)). This is abstraction built on top of `gen.Server` to provide a simple way to create a distributed Producer/Consumer architecture, while automatically managing the concept of backpressure. This implementation is fully compatible with Elixir's GenStage. Example is here [examples/genstage](https://github.com/ergo-services/examples/tree/master/genstage) or just run `go run ./examples/genstage` to see it in action
  * `gen.Saga` behavior support. It implements Saga design pattern - a sequence of transactions that updates each service state and publishes the result (or cancels the transaction or triggers the next transaction step). `gen.Saga` also provides a feature of interim results (can be used as transaction progress or as a part of pipeline processing), time deadline (to limit transaction lifespan), two-phase commit (to make distributed transaction atomic). Here is example [examples/gensaga](https://github.com/ergo-services/examples/tree/master/gensaga).
  * `gen.Raft` behavior support. It's improved implementation of [Raft consensus algorithm](https://raft.github.io). The key improvement is using quorum under the hood to manage the leader election process and make the Raft cluster more reliable. This implementation supports quorums of 3, 5, 7, 9, or 11 quorum members. Here is an example of this feature [examples/genraft](https://github.com/ergo-services/examples/tree/master/genraft)
* Monitor processes/nodes, local/remote with Erlang support
* Link processes local/remote with Erlang support
* [embedded EPMD](#epmd) (in order to get rid of erlang' dependencies) with Erlang support
* Unmarshalling terms into the struct using `etf.TermIntoStruct`, `etf.TermProplistIntoStruct` or to the string using `etf.TermToString` including custom marshaling/unmarshaling via `Marshal` and `Unmarshal` interfaces. But it's highly recommended to use `etf.RegisterType` so you will be receiving messages in a native Golang-type
* Encryption (TLS 1.3) support (including autogenerating self-signed certificates)
* Compression support (with customization of compression level and threshold). It can be configured for the node or a particular process.
* Proxy support with end-to-end encryption, includeing compression/fragmentation/linking/monitoring features.
* Tested and confirmed support Windows, Darwin (MacOS), Linux, FreeBSD.
* Zero dependencies. All features are implemented using the standard Golang library.

### Requirements ###

* Go 1.17.x and above

### Versioning ###

Golang introduced [v2 rule](https://go.dev/blog/v2-go-modules) a while ago to solve complicated dependency issues. We found this solution very controversial and there is still a lot of discussion around it. So, we decided to keep the old way for the versioning, but have to use the git tag with v1 as a major version (due to "v2 rule" restrictions). Since now we use git tag pattern 1.999.XYZ where X - major number, Y - minor, Z - patch version.

### Changelog ###

Here are the changes of latest release. For more details see the [ChangeLog](ChangeLog.md)

#### [v2.2.2](https://github.com/sllt/ergo/releases/tag/v1.999.222) 2023-03-01 [tag version v1.999.222] ####

* Introduced `gen.Pool`. This behavior implements a basic design pattern with a pool of workers. All messages/requests received by the pool process are forwarded to the workers using the "Round Robin" algorithm. The worker process is automatically restarting on termination. See example here [examples/genpool](https://github.com/ergo-services/examples/tree/master/genpool)
* Removed Erlang RPC support. A while ago Erlang has changed the way of handling this kind of request making this feature more similar to the regular `gen.Server`. So, there is no reason to keep supporting it. Use a regular way of messaging instead - `gen.Server`. 
* Fixed issue #130 (`StartType` option in `gen.ApplicationSpec` is ignored for the autostarting applications)
* Fixed issue #143 (incorrect cleaning up the aliases belonging to the terminated process)

#### [v2.2.1](https://github.com/sllt/ergo/releases/tag/v1.999.221) 2023-02-01 [tag version v1.999.221] ####

* Now you can join your services made with Ergo Framework into a single cluster with transparent networking using our **Cloud Overlay Network** where they can connect to each other smoothly, no matter where they run - AWS, Azure or GCP, or anywhere else. All these connections are secured with end-to-end encryption. Read more in this article [https://blog.ergo.services/cloud-overlay-network-3a133d47efe5](https://blog.ergo.services/cloud-overlay-network-3a133d47efe5). Here is an example of this feature in action [examples/cloud](https://github.com/ergo-services/examples/tree/master/cloud)
* `examples` moved to https://github.com/ergo-services/examples
* Added support Erlang OTP/25
* Improved handling `nil` values for the registered types using `etf.RegisterType(...)`
* Improved self-signed certificate generation
* Introduced `ergo.debug` option that enables extended debug information for `lib.Log(...)`/`lib.Warning(...)`
* Fixed `gen.TCP` and `gen.UDP` (missing callbacks)
* Fixed ETF registering type with `etf.Pid`, `etf.Alias` or `etf.Ref` value types
* Fixed Cloud client
* Fixed #117 (incorrect hanshake process finalization)
* Fixed #139 (panic of the gen.Stage partition dispatcher)

### Benchmarks ###

Here is simple EndToEnd test demonstrates performance of messaging subsystem

Hardware: workstation with AMD Ryzen Threadripper 3970X (64) @ 3.700GHz

```
❯❯❯❯ go test -bench=NodeParallel -run=XXX -benchtime=10s
goos: linux
goarch: amd64
pkg: github.com/sllt/ergo/tests
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor
BenchmarkNodeParallel-64                 4738918              2532 ns/op
BenchmarkNodeParallelSingleNode-64      100000000              429.8 ns/op

PASS
ok      github.com/sllt/ergo/tests  29.596s
```

these numbers show almost **500.000 sync requests per second** for the network messaging via localhost and **10.000.000 sync requests per second** for the local messaging (within a node).

#### Compression

This benchmark shows the performance of compression for sending 1MB message between two nodes (via a network).

```
❯❯❯❯ go test -bench=NodeCompression -run=XXX -benchtime=10s
goos: linux

ChangeLog.md
# Changelog
All notable changes to this project will be documented in this file.

This format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

#### [v2.2.2](https://github.com/sllt/ergo/releases/tag/v1.999.222) 2023-03-01 [tag version v1.999.222] ####

* Introduced `gen.Pool`. This behavior implements a basic design pattern with a pool of workers. All messages/requests received by the pool process are forwarded to the workers using the "Round Robin" algorithm. The worker process is automatically restarting on termination. See example here [examples/genpool](https://github.com/ergo-services/examples/tree/master/genpool)
* Removed Erlang RPC support. A while ago Erlang has changed the way of handling this kind of request making this feature more similar to the regular `gen.Server`. So, there is no reason to keep supporting it. Use a regular way of messaging instead - `gen.Server`. 
* Fixed issue #130 (`StartType` option in `gen.ApplicationSpec` is ignored for the autostarting applications)
* Fixed issue #143 (incorrect cleaning up the aliases belonging to the terminated process)

#### [v2.2.1](https://github.com/sllt/ergo/releases/tag/v1.999.221) 2023-01-18 [tag version v1.999.221] ####

* Now you can join your services made with Ergo Framework into a single cluster with transparent networking using our **Cloud Overlay Network** where they can connect to each other smoothly, no matter where they run - AWS, Azure or GCP, or anywhere else. All these connections are secured with end-to-end encryption. Read more in this article [https://https://medium.com/@ergo-services/cloud-overlay-network](https://https://medium.com/@ergo-services/cloud-overlay-network). Here is an example of this feature in action [examples/cloud](https://github.com/ergo-services/examples/tree/master/cloud)
* `examples` moved to https://github.com/ergo-services/examples
* Added support Erlang OTP/25
* Improved handling `nil` values for the registered types using `etf.RegisterType(...)`
* Improved self-signed certificate generation
* Introduced `ergo.debug` option that enables extended debug information for `lib.Log(...)`/`lib.Warning(...)`
* Fixed `gen.TCP` and `gen.UDP` (missing callbacks)
* Fixed ETF registering type with `etf.Pid`, `etf.Alias` or `etf.Ref` value types
* Fixed Cloud client
* Fixed #117 (incorrect hanshake process finalization)
* Fixed #139 (panic of the gen.Stage partition dispatcher)

#### [v2.2.0](https://github.com/sllt/ergo/releases/tag/v1.999.220) 2022-10-18 [tag version v1.999.220] ####

* Introduced `gen.Web` behavior. It implements **Web API Gateway pattern** is also sometimes known as the "Backend For Frontend" (BFF). See example [examples/genweb](https://github.com/ergo-services/examples/tree/master/genweb)
* Introduced `gen.TCP` behavior - **socket acceptor pool for TCP protocols**. It provides everything you need to accept TCP connections and process packets with a small code base and low latency. Here is simple example [examples/gentcp](https://github.com/ergo-services/examples/tree/master/gentcp)
* Introduced `gen.UDP` - the same as `gen.TCP`, but for UDP protocols. Example is here [examples/genudp](https://github.com/ergo-services/examples/tree/master/genudp)
* Introduced **Events**. This is a simple pub/sub feature within a node - any `gen.Process` can become a producer by registering a new event `gen.Event` using method `gen.Process.RegisterEvent`, while the others can subscribe to these events using `gen.Process.MonitorEvent`. Subscriber process will also receive `gen.MessageEventDown` if a producer process went down (terminated). This feature behaves in a monitor manner but only works within a node. You may also want to subscribe to a system event - `node.EventNetwork` to receive event notification on connect/disconnect any peers.
* Introduced **Cloud Client** - allows connecting to the cloud platform [https://ergo.sevices](https://ergo.services). You may want to register your email there, and we will inform you about the platform launch day
* Introduced **type registration** for the ETF encoding/decoding. This feature allows you to get rid of manually decoding with `etf.TermIntoStruct` for the receiving messages. Register your type using `etf.RegisterType(...)`, and you will be receiving messages in a native type
* Predefined set of errors has moved to the `lib` package
* Updated `gen.ServerBehavior.HandleDirect` method (got extra argument `etf.Ref` to distinguish the requests). This change allows you to handle these requests asynchronously using method `gen.ServerProcess.Reply(...)`
* Updated `node.Options`. Now it has field `Listeners` (type `node.Listener`). It allows you to start any number of listeners with custom options - `Port`, `TLS` settings, or custom `Handshake`/`Proto` interfaces
* Fixed build on 32-bit arch
* Fixed freezing on ARM arch #102
* Fixed problem with encoding negative int8
* Fixed #103 (there was an issue on interop with Elixir's GenStage)
* Fixed node stuck on start if it uses the name which is already taken in EPMD

etf/etf.go
package etf

import (
	"fmt"
	"hash/crc32"
	"reflect"
	"strings"
	"sync"

	"github.com/sllt/ergo/lib"
)

var (
	registered = registeredTypes{
		typesEnc: make(map[Atom]*registerType),
		typesDec: make(map[Atom]*registerType),
	}
)

// Erlang external term tags.
const (
	ettAtom          = byte(100) //deprecated
	ettAtomUTF8      = byte(118)
	ettSmallAtom     = byte(115) //deprecated
	ettSmallAtomUTF8 = byte(119)
	ettString        = byte(107)

	ettCacheRef = byte(82)

	ettNewFloat = byte(70)

	ettSmallInteger = byte(97)
	ettInteger      = byte(98)
	ettLargeBig     = byte(111)
	ettSmallBig     = byte(110)

	ettList         = byte(108)
	ettListImproper = byte(18) // to be able to encode improper lists like [a|b].
	ettSmallTuple   = byte(104)
	ettLargeTuple   = byte(105)

	ettMap = byte(116)

	ettBinary    = byte(109)
	ettBitBinary = byte(77)

	ettNil = byte(106)

	ettPid      = byte(103)
	ettNewPid   = byte(88) // since OTP 23, only when BIG_CREATION flag is set
	ettNewRef   = byte(114)
	ettNewerRef = byte(90) // since OTP 21, only when BIG_CREATION flag is set

	ettExport = byte(113)
	ettFun    = byte(117) // legacy, wont support it here
	ettNewFun = byte(112)

	ettPort    = byte(102)
	ettNewPort = byte(89) // since OTP 23, only when BIG_CREATION flag is set

	// ettRef        = byte(101) deprecated

	ettFloat = byte(99) // legacy
)

type registeredTypes struct {
	sync.RWMutex
	typesEnc map[Atom]*registerType
	typesDec map[Atom]*registerType
}
type registerType struct {
	rtype  reflect.Type
	name   Atom
	origin Atom
	strict bool
}

// Term
type Term interface{}

// Tuple
type Tuple []Term

// List
type List []Term

// Alias
type Alias Ref

// ListImproper as a workaround for the Erlang's improper list [a|b]. Intended to be used to interact with Erlang.
type ListImproper []Term

// Atom
type Atom string

// Map
type Map map[Term]Term

// String this type is intended to be used to interact with Erlang. String value encodes as a binary (Erlang type: <<...>>)
type String string

// Charlist this type is intended to be used to interact with Erlang. Charlist value encodes as a list of int32 numbers in order to support Erlang string with UTF-8 symbols on an Erlang side (Erlang type: [...])
type Charlist string

// Pid
type Pid struct {
	Node     Atom
	ID       uint64
	Creation uint32
}

// Port
type Port struct {
	Node     Atom
	ID       uint32
	Creation uint32
}

// Ref
type Ref struct {
	Node     Atom
	Creation uint32
	ID       [5]uint32
}

// Marshaler interface implemented by types that can marshal themselves into valid ETF binary
// Interface implementation must be over the object e.g. (MyObject) UnmarshalETF:
//
//		type MyObject struct{}
//
//		func (m MyObject) MarshalETF() ([]byte, error) {
//			var encoded []byte
//			... encoding routine ...
//			return encoded, nil
//	}
type Marshaler interface {
	MarshalETF() ([]byte, error)
}

// Unmarshaler interface implemented by types that can unmarshal an ETF binary of themselves.
// Returns error ErrEmpty for []byte{}.
// Interface implementation must be over pointer to the object e.g. (*MyObject) UnmarshalETF:
//
//	type MyObject struct{}
//
//	func (m *MyObject) UnmarshalETF(b []byte) error {
//		var err error
//		... decoding routine ...
//		return err
//	}
type Unmarshaler interface {
	UnmarshalETF([]byte) error
}

// Function
type Function struct {
	Arity  byte
	Unique [16]byte
	Index  uint32
	//	Free      uint32
	Module    Atom
	OldIndex  uint32
	OldUnique uint32
	Pid       Pid
	FreeVars  []Term
}

// Export
type Export struct {
	Module   Atom
	Function Atom
	Arity    int
}

// Element
func (m Map) Element(k Term) Term {
	return m[k]
}

// Element
func (l List) Element(i int) Term {
	return l[i-1]
}

// Element
func (t Tuple) Element(i int) Term {
	return t[i-1]
}

// String
func (p Pid) String() string {
	empty := Pid{}
	if p == empty {
		return "<0.0.0>"
	}

	n := uint32(0)
	if p.Node != "" {
		n = crc32.Checksum([]byte(p.Node), lib.CRC32Q)
	}
	return fmt.Sprintf("<%08X.%d.%d>", n, int32(p.ID>>32), int32(p.ID))
}

// String
func (r Ref) String() string {
	n := uint32(0)
	if r.Node != "" {
		n = crc32.Checksum([]byte(r.Node), lib.CRC32Q)
	}
	return fmt.Sprintf("Ref#<%08X.%d.%d.%d>", n, r.ID[0], r.ID[1], r.ID[2])
}

// String
func (a Alias) String() string {
	n := uint32(0)
	if a.Node != "" {
		n = crc32.Checksum([]byte(a.Node), lib.CRC32Q)
	}
	return fmt.Sprintf("Ref#<%08X.%d.%d.%d>", n, a.ID[0], a.ID[1], a.ID[2])
}

// ProplistElement
type ProplistElement struct {
	Name  Atom
	Value Term
}

// TermToString transforms given term (Atom, []byte, List) to the string
func TermToString(t Term) (s string, ok bool) {
	ok = true
	switch x := t.(type) {
	case Atom:
		s = string(x)
	case string:
		s = x
	case []byte:
		s = string(x)
	case List:
		str, err := convertCharlistToString(x)
		if err != nil {
			ok = false
			return
		}
		s = str
	default:
		ok = false
	}
	return
}

// TermProplistIntoStruct transorms given term into the provided struct 'dest'.
// Proplist is the list of Tuple values with two items { Name , Value },
// where Name can be string or Atom and Value must be the same type as
// it has the field of 'dest' struct with the equivalent name. Its also
// accepts []ProplistElement as a 'term' value
func TermProplistIntoStruct(term Term, dest interface{}) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("%v", r)
		}
	}()
	v := reflect.Indirect(reflect.ValueOf(dest))
	return setProplist(term, v)
}

// TermIntoStruct transforms 'term' (etf.Term, etf.List, etf.Tuple, etf.Map) into the
// given 'dest' (could be a struct, map, slice or array). Its a pretty
// expencive operation in terms of CPU usage so you shouldn't use it
// on highload parts of your code. Use manual type casting instead.
func TermIntoStruct(term Term, dest interface{}) (err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("%v", r)
		}
	}()
	v := reflect.Indirect(reflect.ValueOf(dest))
	err = termIntoStruct(term, v)
	return
}

func termIntoStruct(term Term, dest reflect.Value) error {

	if term == nil {
		return nil
	}

	if dest.Type().NumMethod() > 0 && dest.CanInterface() {
		v := dest
		if v.Kind() != reflect.Ptr && v.CanAddr() {
			v = v.Addr()

			if u, ok := v.Interface().(Unmarshaler); ok {
				b, is_binary := term.([]byte)
				if !is_binary {
					return fmt.Errorf("can't unmarshal value, wront type %s", term)
				}
				return u.UnmarshalETF(b)
			}
		}
	}

	switch dest.Kind() {
	case reflect.Ptr:
		pdest := reflect.New(dest.Type().Elem())
		dest.Set(pdest)
		dest = pdest.Elem()
		return termIntoStruct(term, dest)

	case reflect.Array, reflect.Slice:
		t := dest.Type()
		byte_slice, ok := term.([]byte)
		if t == reflect.SliceOf(reflect.TypeOf(byte(1))) && ok {
			dest.Set(reflect.ValueOf(byte_slice))
			return nil

		}
		if _, ok := term.(List); !ok {
			// in case if term is the golang native type
			dest.Set(reflect.ValueOf(term))
			return nil
		}
		return setListField(term.(List), dest)

	case reflect.Struct:
		switch s := term.(type) {
		case Map:
			return setMapStructField(s, dest)
		case Tuple:
			return setStructField(s, dest)
		case Ref:
			dest.Set(reflect.ValueOf(s))
			return nil
		case Pid:
			dest.Set(reflect.ValueOf(s))
			return nil
		}
		return fmt.Errorf("can't convert %#v to struct", term)

	case reflect.Map:
		if _, ok := term.(Map); !ok {
			// in case if term is the golang native type
			dest.Set(reflect.ValueOf(term))
			return nil
		}
		return setMapField(term.(Map), dest)

	case reflect.Bool:
		b, ok := term.(bool)
		if !ok {
			return fmt.Errorf("can't convert %#v to bool", term)
		}
		dest.SetBool(b)
		return nil

	case reflect.Float32, reflect.Float64:
		f, ok := term.(float64)
		if !ok {
			return fmt.Errorf("can't convert %#v to float64", term)
		}
		dest.SetFloat(f)
		return nil

	case reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64, reflect.Int:
		i := int64(0)
		switch v := term.(type) {
		case int64:
			i = v
		case int32:
			i = int64(v)
		case int16:
			i = int64(v)
		case int8:
			i = int64(v)
		case int:
			i = int64(v)
		case uint64:
			i = int64(v)
		case uint32:
			i = int64(v)
		case uint16:
			i = int64(v)
		case uint8:
			i = int64(v)
		case uint:
			i = int64(v)
		default:
			return fmt.Errorf("can't convert %#v to int64", term)
		}
		dest.SetInt(i)
		return nil

	case reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uint:
		u := uint64(0)
		switch v := term.(type) {
		case uint64:
			u = v
		case uint32:
			u = uint64(v)
		case uint16:
			u = uint64(v)
		case uint8:
			u = uint64(v)
		case uint:
			u = uint64(v)
		case int64:
			u = uint64(v)
		case int32:
			u = uint64(v)
		case int16:
			u = uint64(v)
		case int8:
			u = uint64(v)
		case int:
			u = uint64(v)

		default:
			return fmt.Errorf("can't convert %#v to uint64", term)
		}
		dest.SetUint(u)
		return nil

	case reflect.String:
		switch v := term.(type) {
		case List:
			s, err := convertCharlistToString(v)
			if err != nil {
				return err
			}
			dest.SetString(s)
			return nil
		case []byte:
			dest.SetString(string(v))
			return nil
		case string:
			dest.SetString(v)
			return nil
		case Atom:
			dest.SetString(string(v))
			return nil
		}

	default:
		dest.Set(reflect.ValueOf(term))
		return nil
	}

	return nil
}

func setListField(term List, dest reflect.Value) error {
	var value reflect.Value
	if dest.Kind() == reflect.Ptr {
		pdest := reflect.New(dest.Type().Elem())
		dest.Set(pdest)
		dest = pdest.Elem()
	}
	t := dest.Type()
	switch t.Kind() {
	case reflect.Slice:
		value = reflect.MakeSlice(t, len(term), len(term))
	case reflect.Array:
		if t.Len() != len(term) {
			return NewInvalidTypesError(t, term)
		}
		value = dest
	default:
		return NewInvalidTypesError(t, term)
	}

	for i, elem := range term {
		if err := termIntoStruct(elem, value.Index(i)); err != nil {
			return err
		}
	}

	if t.Kind() == reflect.Slice {
		dest.Set(value)
	}

	return nil
}

func setProplist(term Term, dest reflect.Value) error {
	switch v := term.(type) {
	case []ProplistElement:
		return setProplistElementField(v, dest)
	case List:
		return setProplistField(v, dest)
	default:
		return NewInvalidTypesError(dest.Type(), term)
	}

}

func setProplistField(list List, dest reflect.Value) error {
	t := dest.Type()
	numField := t.NumField()
	fields := make([]reflect.StructField, numField)
	for i := range fields {
		fields[i] = t.Field(i)
	}

	for _, elem := range list {
		if len(elem.(Tuple)) != 2 {
			return &InvalidStructKeyError{Term: elem}
		}

		key := elem.(Tuple)[0]
		val := elem.(Tuple)[1]
		fName, ok := TermToString(key)
		if !ok {
			return &InvalidStructKeyError{Term: key}
		}
		index := findStructField(fields, fName)
		if index == -1 {
			continue
		}

		err := termIntoStruct(val, dest.Field(index))
		if err != nil {
			return err
		}
	}

	return nil
}

func setProplistElementField(proplist []ProplistElement, dest reflect.Value) error {
	t := dest.Type()
	numField := t.NumField()
	fields := make([]reflect.StructField, numField)
	for i := range fields {
		fields[i] = t.Field(i)
	}

	for _, elem := range proplist {
		fName, ok := TermToString(elem.Name)
		if !ok {
			return &InvalidStructKeyError{Term: elem.Name}
		}
		index := findStructField(fields, fName)
		if index == -1 {
			continue
		}

		err := termIntoStruct(elem.Value, dest.Field(index))
		if err != nil {
			return err
		}
	}

	return nil
}
func setMapField(term Map, dest reflect.Value) error {
	switch dest.Type().Kind() {
	case reflect.Map:
		return setMapMapField(term, dest)
	case reflect.Struct:
		return setMapStructField(term, dest)
	case reflect.Interface:
		dest.Set(reflect.ValueOf(term))
		return nil
	}

	return NewInvalidTypesError(dest.Type(), term)
}

func setStructField(term Tuple, dest reflect.Value) error {
	if dest.Kind() == reflect.Ptr {
		pdest := reflect.New(dest.Type().Elem())
		dest.Set(pdest)
		dest = pdest.Elem()
	}
	for i, elem := range term {
		// let it panic if number of term elements is bigger than
		// number of struct fields
		if err := termIntoStruct(elem, dest.Field(i)); err != nil {
			return err
		}
	}

	return nil

}

func setMapStructField(term Map, dest reflect.Value) error {
	t := dest.Type()
	numField := t.NumField()
	fields := make([]reflect.StructField, numField)
	for i := range fields {
		fields[i] = t.Field(i)
	}

	for key, val := range term {
		fName, ok := TermToString(key)
		if !ok {
			return &InvalidStructKeyError{Term: key}
		}
		index := findStructField(fields, fName)
		if index == -1 {
			continue
		}

		err := termIntoStruct(val, dest.Field(index))
		if err != nil {
			return err
		}
	}

	return nil
}

func findStructField(term []reflect.StructField, key string) (index int) {
	var fieldName string
	index = -1
	for i, f := range term {
		fieldName = f.Name

		if tag := f.Tag.Get("etf"); tag != "" {
			fieldName = tag
		}

		if fieldName == key {
			index = i
			return
		} else {
			if strings.EqualFold(f.Name, key) {
				index = i
			}
		}
	}

	return
}

func setMapMapField(term Map, dest reflect.Value) error {
	t := dest.Type()
	if dest.IsNil() {
		dest.Set(reflect.MakeMapWithSize(t, len(term)))
	}
	tkey := t.Key()
	tval := t.Elem()
	for key, val := range term {
		destkey := reflect.Indirect(reflect.New(tkey))
		if err := termIntoStruct(key, destkey); err != nil {
			return err
		}
		destval := reflect.Indirect(reflect.New(tval))
		if err := termIntoStruct(val, destval); err != nil {
			return err
		}
		dest.SetMapIndex(destkey, destval)
	}
	return nil
}

// RegisterTypeOptins defines custom name for the registering type.
// Leaving the Name option empty makes the name automatically generated.
// Strict option defines whether the decoding process causes panic
// if the decoding value doesn't fit the destination object.
type RegisterTypeOptions struct {
	Name   Atom
	Strict bool
}

// RegisterType registers new type with the given options. It returns a Name
// of the registered type, which can be used in the UnregisterType function
// for unregistering this type.
// Returns an error if this type can not be registered.
func RegisterType(t interface{}, options RegisterTypeOptions) (Atom, error) {
	switch t.(type) {
	case Pid, Ref, Alias:
		return "", fmt.Errorf("types Pid, Ref, Alias can not be registered")
	}
	tt := reflect.TypeOf(t)
	ttk := tt.Kind()

	name := options.Name
	origin := regTypeName(tt)
	if name == "" {
		name = origin
	}
	lname := len([]rune(name))
	if lname > 255 {
		return name, fmt.Errorf("type name %q is too long. characters number %d (limit: 255)", name, lname)
	}

	switch ttk {
	case reflect.Struct, reflect.Slice, reflect.Array:
	case reflect.Map:
		// Using pointers for the network messaging is meaningless.
		// Supporting this feature in the maps is getting the decoding process a bit overloaded.
		// But they still can be used for the other types, even being meaningless.
		if tt.Key().Kind() == reflect.Ptr {
			return name, fmt.Errorf("pointer as a key for the map is not supported")
		}
		if tt.Elem().Kind() == reflect.Ptr {
			return name, fmt.Errorf("pointer as a value for the map is not supported")
		}
		// supported types
	default:
		return name, fmt.Errorf("type %q is not supported", regTypeName(tt))
	}

	registered.Lock()
	defer registered.Unlock()

	_, taken := registered.typesDec[name]
	if taken {
		return name, lib.ErrTaken
	}

	r, taken := registered.typesEnc[origin]
	if taken {
		return name, fmt.Errorf("type is already registered as %q", r.name)
	}

	checkIsRegistered := func(name Atom, rt reflect.Kind) error {
		switch rt {
		case reflect.Struct, reflect.Array, reflect.Slice, reflect.Map:
			// check if this type is registered
			_, taken := registered.typesEnc[name]
			if taken == false {
				return fmt.Errorf("type %q must be registered first", name)
			}
		case reflect.Chan, reflect.Func, reflect.UnsafePointer, reflect.Complex64, reflect.Complex128:
			return fmt.Errorf("type %q is not supported", rt)
		}
		return nil
	}

	switch ttk {
	case reflect.Struct:
		// check for unexported fields
		tv := reflect.ValueOf(t)
		for i := 0; i < tv.NumField(); i++ {
			f := tv.Field(i)
			if f.CanInterface() == false {
				return name, fmt.Errorf("struct has unexported field(s)")
			}

			switch f.Interface().(type) {
			case Pid, Ref, Alias:
				// ignore this types
				continue
			}

			if f.Type().Kind() == reflect.Slice && f.Type().Elem().Kind() == reflect.Uint8 {
				// []byte
				continue
			}

			orig := regTypeName(f.Type())
			if err := checkIsRegistered(orig, f.Kind()); err != nil {
				return name, err
			}
		}
	case reflect.Array, reflect.Slice, reflect.Map:
		elem := tt.Elem()
		orig := regTypeName(elem)
		if err := checkIsRegistered(orig, elem.Kind()); err != nil {
			return name, err
		}
	}

	rt := &registerType{
		rtype:  reflect.TypeOf(t),
		name:   name,
		origin: origin,
		strict: options.Strict,
	}
	registered.typesEnc[origin] = rt
	registered.typesDec[name] = rt
	return name, nil
}

// UnregisterType unregisters type with a given name.
func UnregisterType(name Atom) error {
	registered.Lock()
	defer registered.Unlock()
	r, found := registered.typesDec[name]
	if found == false {
		return lib.ErrUnknown
	}
	delete(registered.typesDec, name)
	delete(registered.typesEnc, r.origin)
	return nil
}

type StructPopulatorError struct {
	Type reflect.Type
	Term Term
}

func (s *StructPopulatorError) Error() string {
	return fmt.Sprintf("Cannot put %#v into go value of type %s", s.Term, s.Type.Kind().String())
}

func NewInvalidTypesError(t reflect.Type, term Term) error {
	return &StructPopulatorError{
		Type: t,
		Term: term,
	}
}

type InvalidStructKeyError struct {
	Term Term
}

func (s *InvalidStructKeyError) Error() string {
	return fmt.Sprintf("Cannot use %s as struct field name", reflect.TypeOf(s.Term).Name())
}

func convertCharlistToString(l List) (string, error) {
	runes := make([]rune, len(l))
	for i := range l {
		switch x := l[i].(type) {
		case int64:
			runes[i] = int32(x)
		case int32:
			runes[i] = int32(x)
		case int16:
			runes[i] = int32(x)
		case int8:
			runes[i] = int32(x)
		case int:
			runes[i] = int32(x)
		default:
			return "", fmt.Errorf("wrong rune %#v", l[i])
		}
	}
	return string(runes), nil
}

func regTypeName(t reflect.Type) Atom {
	return Atom("#" + t.PkgPath() + "/" + t.Name())
}
